gh-101881: os: support blocking functions on Windows

RayyanAnsari opened the pull request:
Use the GetNamedPipeHandleState and SetNamedPipeHandleState Win32 API functions to add support for os.get_blocking and os.set_blocking.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-101881 -->
* Issue: gh-101881
<!-- /gh-issue-number -->

RayyanAnsari committed:
fileutils: handle non-blocking pipe IO on Windows

Handle erroring operations on non-blocking pipes by reading the _doserrno code.
Limit writes on non-blocking pipes that are too large.

eryksun commented on Python/fileutils.c:
```diff
@@ -1750,7 +1750,16 @@ _Py_read(int fd, void *buf, size_t count)
+                           returns EINVAL on Windows */
+                        errno = EAGAIN;
+                }
+        }
```
Per PEP 7, a comment should go before the code that it describes; the C99 `//` comment style is allowed; and indents should be 4 spaces. Also, I think it would be better to clarify the context in the comment instead of adding the redundant variable `winerr`.

```suggestion
        // read() on a non-blocking empty pipe fails with EINVAL, which is
        // mapped from the Windows error code ERROR_NO_DATA.
        if (n < 0 && errno == EINVAL) {
            if (_doserrno == ERROR_NO_DATA) {
                errno = EAGAIN;
            }
        }
```

eryksun commented on Python/fileutils.c:
```diff
@@ -1814,7 +1823,14 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+                /* write() on a non-blocking full pipe
+                   returns ENOSPC on Windows */
+                errno = EAGAIN;
+            }
```
This case isn't limited to just a full pipe. There are no partial writes up to the available space in non-blocking mode, contrary to the documentation. I tested this as far back as Windows NT 4.0. The official WinAPI documentation is wrong.

```suggestion
            _doserrno = 0;
            n = write(fd, buf, (int)count);
            // write() on a non-blocking pipe fails with ENOSPC on Windows if
            // the pipe lacks available space for the entire buffer.
            if (n < 0 && errno == ENOSPC && _doserrno == 0) {
                errno = EAGAIN;
            }
```

Edit: actually `_doserrno` is needed to clarify that `WriteFile()` succeeded with 0 bytes written.

eryksun commented on Python/fileutils.c:
```diff
@@ -1829,7 +1845,14 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+                /* write() on a non-blocking full pipe
+                   returns ENOSPC on Windows */
+                errno = EAGAIN;
+            }
```
```suggestion
            _doserrno = 0;
            n = write(fd, buf, (int)count);
            // write() on a non-blocking pipe fails with ENOSPC on Windows if
            // the pipe lacks available space for the entire buffer.
            if (n < 0 && errno == ENOSPC && _doserrno == 0) {
                errno = EAGAIN;
            }
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE)
+        return -1;
```
```suggestion
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
+    if (handle == INVALID_HANDLE_VALUE)
+        return -1;
+
+    if (!GetNamedPipeHandleState(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
```
We link directly to [W]ide-character functions instead of relying on the macros. It's good to be consistent about this even though it doesn't technically matter here.

```suggestion
    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    return mode == PIPE_WAIT;
```
The value could include `PIPE_READMODE_MESSAGE` (2) even though it's not something we directly support.

```suggestion
    return mode & PIPE_WAIT;
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE)
+        return -1;
```
```suggestion
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2473,41 @@ _Py_set_blocking(int fd, int blocking)
+    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
```
`PIPE_READMODE_MESSAGE` needs to be retained if it's enabled.

```suggestion
    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
    mode |= blocking ? PIPE_WAIT : PIPE_NOWAIT;
    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
```

RayyanAnsari committed:
gh-101881: os: support blocking functions on Windows

Use the GetNamedPipeHandleState and SetNamedPipeHandleState Win32 API functions to add support for os.get_blocking and os.set_blocking.

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2470,47 @@ _Py_set_blocking(int fd, int blocking)
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    return mode & PIPE_WAIT;
```
The GIL has to be released here. Also, I messed up the return value. It has to test for `PIPE_NOWAIT` (1) and invert the result.

```suggestion
    DWORD mode;
    BOOL success;

    handle = _Py_get_osfhandle(fd);
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }

    Py_BEGIN_ALLOW_THREADS
    success = GetNamedPipeHandleStateW(handle, &mode,
                                       NULL, NULL, NULL, NULL, 0);
    Py_END_ALLOW_THREADS
    
    if (!success) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
    
    return !(mode & PIPE_NOWAIT);
```

The GIL has to be release because get/set pipe operations use the same synchronization lock that's used by synchronous reads and writes. Thus these operations can block if another thread is waiting for I/O. For example, using the code from this pull request:

```pycon
>>> import os, threading
>>> r, w = os.pipe()
>>> threading.Thread(target=os.read, args=(r, 1)).start()
>>> os.get_blocking(r) # This blocks.
```

eryksun commented on Python/fileutils.c:
```diff
@@ -2450,6 +2470,47 @@ _Py_set_blocking(int fd, int blocking)
+    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
```
Same for the GIL here, and my previous suggestion messed up the logic here as well. The `PIPE_NOWAIT` flag has to be cleared when `blocking` is true.

```suggestion
    DWORD mode;
    BOOL success;

    handle = _Py_get_osfhandle(fd);
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }

    Py_BEGIN_ALLOW_THREADS
    success = GetNamedPipeHandleStateW(handle, &mode,
                                       NULL, NULL, NULL, NULL, 0);
    if (success) {
        if (blocking) {
            mode &= ~PIPE_NOWAIT;
        }
        else {
            mode |= PIPE_NOWAIT;
        }
        success = SetNamedPipeHandleState(handle, &mode, NULL, NULL);
    }
    Py_END_ALLOW_THREADS

    if (!success) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
```

eryksun commented on Python/fileutils.c:
```diff
@@ -1804,6 +1812,41 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+                count = pipe_size;
+            }
+        }
+    }
```
We have to release the GIL when calling `GetFiletype()` and `GetNamedPipeHandleStateW()`.

```suggestion
    if (gil_held) {
        Py_BEGIN_ALLOW_THREADS
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
                pipe_size = 4096;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
        Py_END_ALLOW_THREADS
    }
    else {
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
                pipe_size = 4096;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
    }
```
---

`GetNamedPipeHandleStateW()` can block because it's a synchronous operation that has to acquire the file lock in the kernel. `GetFiletype()` is normally fast and doesn't block as long as the real device object is directly accessible from the kernel file object. However, even `GetFileType()` can block if the file is opened on the MUP device (multiple UNC provider) and redirected through a provider such as SMB. For example, "//./pipe/pipename" is not redirected since "//./" is the local device namespace. OTOH, "//server/pipe/pipename" opens on the MUP device and is redirected through the SMB provider, even if "server" is the local machine name.

RayyanAnsari replied:
The concept of the GIL is new to me, so thanks for the feedback here.

eryksun commented on Python/fileutils.c:
```diff
@@ -1804,6 +1812,50 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+                count = pipe_size;
+            }
+        }
+    }
```
I overlooked outbound and duplex pipes, which allow writes to the server end of the pipe. Writing to the server end is limited to the out-buffer size/quota. The `GetNamedPipeInfo()` call needs to get the pipe flags. If `PIPE_SERVER_END` is set, use the out-buffer size, else use the in-buffer size.

```suggestion
    DWORD pipe_mode, pipe_flags, pipe_size, pipe_insize, pipe_outsize;
    if (gil_held) {
        Py_BEGIN_ALLOW_THREADS
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            pipe_mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
                                  &pipe_insize, NULL))
            {
                pipe_size = 4096;
            }
            else {
                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
                                pipe_insize;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
        Py_END_ALLOW_THREADS
    }
    else {
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &pipe_mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            pipe_mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, &pipe_flags, &pipe_outsize,
                                  &pipe_insize, NULL))
            {
                pipe_size = 4096;
            }
            else {
                pipe_size = pipe_flags & PIPE_SERVER_END ? pipe_outsize :
                                pipe_insize;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
    }
```

zooba commented on Doc/library/os.rst:
```diff
@@ -1091,13 +1091,17 @@ as internal buffering of data.
    .. versionadded:: 3.5
 
+   .. versionchanged:: 3.12
+      Added support for Windows.
```
```suggestion
      Added support for pipes on Windows.
```

zooba commented on Include/internal/pycore_fileutils.h:
```diff
@@ -160,11 +160,10 @@ PyAPI_FUNC(int) _Py_set_inheritable_async_safe(int fd, int inheritable,
 
 PyAPI_FUNC(int) _Py_set_blocking(int fd, int blocking);
-#else   /* MS_WINDOWS */
+#ifdef MS_WINDOWS
```
```suggestion

#ifdef MS_WINDOWS
```

zooba commented on Python/fileutils.c:
```diff
@@ -1804,6 +1812,62 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+
+    // In non-blocking mode, a write that exceeds the available size of a pipe
+    // fails without writing anything. Limiting writes to the pipe size allows
+    // a buffered write to succeed eventually, as the pipe is read.
```
Can we move this work until after we receive that error?

I'd prefer not to penalise users who stay below the buffer limit by making them do all this extra validation each time.

zooba replied:
At a minimum, can we refactor the main check out into its own function so that someone reading this bit can just see the name of what it's doing and not have to work through the logic twice.

eryksun replied:
@zooba, if we retry the write, what do you think about iterating with half the previous count value until it reaches 1?

zooba replied:
It seems like we won't get any buffer size info back from the call, so that's really the only algorithm that makes sense. Would be great if we could raise a warning as well, but we don't really have any precedent for that. People can track on their own if they're regularly getting `n < len(data)` back from their write call.

eryksun replied:
> It seems like we won't get any buffer size info back from the call, so that's really the only algorithm that makes sense. 

For a nonblocking pipe, a `WriteFile()` call that exceeds the available space in the pipe 'succeeds' with 0 bytes written. C `write()` handles this by failing with `ENOSPC`. AFAIK, the Windows API doesn't provide a way to query the available write quota. The internal query that's used by `GetNamedPipeInfo()` actually does return the available write quota, but the Windows API ignores it. 

Currently this PR caps the write at the pipe size, even if it's not all available. Eventually the reader will empty the pipe, and the full write will succeed. This works, albeit less aggressively than trying the write multiple times with a decreasing size until it succeeds. The latter is closer in overall behavior to Linux.

eryksun replied:
Here's a simple example of the behavior when the write is retried in a loop with a decreasing count. Note that `CreatePipe()` uses a default pipe size of 4 KiB.

```pycon
>>> import os
>>> r, w = os.pipe()
>>> os.set_blocking(w, False)
>>> os.write(w, b'a' * 3072)
3072
>>> os.write(w, b'a' * 2048)
1024
>>> os.write(w, b'a' * 2048)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
BlockingIOError: [Errno 11] Resource temporarily unavailable
>>> os.read(r, 4096) == b'a' * 4096
True
```

zooba commented on Doc/library/os.rst:
```diff
@@ -1091,13 +1091,17 @@ as internal buffering of data.
       The function is limited on Emscripten and WASI, see
       :ref:`wasm-availability` for more information.
 
+      On Windows, this function is limited to pipe files.
```
```suggestion
      On Windows, this function is limited to pipes.
```

zooba commented on Doc/library/os.rst:
```diff
@@ -1565,13 +1569,17 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
       The function is limited on Emscripten and WASI, see
       :ref:`wasm-availability` for more information.
 
+      On Windows, this function is limited to pipe files.
```
```suggestion
      On Windows, this function is limited to pipes.
```

zooba commented on Doc/library/os.rst:
```diff
@@ -1565,13 +1569,17 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
    .. versionadded:: 3.5
 
+   .. versionchanged:: 3.12
+      Added support for Windows.
```
```suggestion
      Added support for pipes on Windows.
```

eryksun commented on Python/fileutils.c:
```diff
@@ -1804,6 +1812,62 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+                count = pipe_size;
+            }
+        }
+    }
```
As discussed, retry using a loop and decreasing count instead of capping the write at the pipe size.

```suggestion
```

eryksun commented on Python/fileutils.c:
```diff
@@ -1814,7 +1878,13 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+            // the pipe lacks available space for the entire buffer.
+            if (n < 0 && errno == ENOSPC && _doserrno == 0) {
+                errno = EAGAIN;
+            }
```
Here's a suggestion for the retry loop.

```suggestion
            // write() on a non-blocking pipe fails with ENOSPC on Windows if
            // the pipe lacks available space for the entire buffer.
            int c = (int)count;
            do {
                _doserrno = 0;
                n = write(fd, buf, c);
                if (n >= 0 || errno != ENOSPC || _doserrno != 0) {
                    break;
                }
                errno = EAGAIN;
                c /= 2;
            } while (c > 0);
```

eryksun commented on Python/fileutils.c:
```diff
@@ -1829,7 +1899,13 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
+            // the pipe lacks available space for the entire buffer.
+            if (n < 0 && errno == ENOSPC && _doserrno == 0) {
+                errno = EAGAIN;
+            }
```
```suggestion
            // write() on a non-blocking pipe fails with ENOSPC on Windows if
            // the pipe lacks available space for the entire buffer.
            int c = (int)count;
            do {
                _doserrno = 0;
                n = write(fd, buf, c);
                if (n >= 0 || errno != ENOSPC || _doserrno != 0) {
                    break;
                }
                errno = EAGAIN;
                c /= 2;
            } while (c > 0);
```
